gco rans3p

Giovanni-Cozzuto-1989 opened the pull request:
This PR is for working with the three phase solver and creating benchmark for sediment transport problems.

- [x] Conservation is preserved.
- [x] Check and fix hsu stress formulation in SedClosure.
- [x] Check and fix the signs of momentum interface for both fluid and sediment.
- [x] Check and fix the signs of hamiltonian term (pressure) for both fluid and sediment.
- [x] Check and fix the use of phi_s in RANS3PF and RANS3PSed coupled with vos.
- [x] Check and fix the use of density through sediment and fluid model.
- [x] Check and fix the penalty coefficient used for perturbating Navier-Stokes.
- [x] Check and fix the post-step correction terms in velocity field for both sediment and fluid phases.
- [x] Check and fix both density and viscosity used for the momentum interface calculation in SedClosure.
- [x] Check and fix sediment pressure for frictional contact.
- [x] Check and fix sediment shear stress for frictional contact. 
- [x] Check and fix limiters to prevent pressure blowing up.  
- [x] Check and fix gradient of vos to be plugged into the pressure correction in contact stress module.

Giovanni-Cozzuto-1989 commented:
@cekees @manuel-quezada 
At some point RANS3PSed was missing ebqe_velocity_last in its Coefficients class.
Does this fix look right?
https://github.com/erdc/proteus/blob/5d0eb7046ef4bfbe16e8aacdd48632371411d4ef/proteus/mprans/RANS3PSed.py#L631

cekees commented:
Yes, that looks right on the ebqe_velocity_last issue. Either @manuel-quezada or I should do a comparison of RANS3PF and RANS3PSed fairly soon since they are have closely related numerics.

cekees commented:
This all looks good except for the setup.py file. It seems to undo recent changes. I will check it more closely.

Giovanni-Cozzuto-1989 commented:
Yeah I know, after that I locally updated the stack and setup.py according to the changes @tridelat made few weeks ago.
However, I'm still working on this branh, please don't merge it yet.

adimako commented:
@cekees , @Giovanni-Cozzuto-1989 showed me an initial result. Conservation for the solid is ok, so this is a good start
We found also some bugs in the set-up files @Giovanni-Cozzuto-1989 will fix these and re-run to see if the sediment is falling (the bug was that rho_f=rho_s, so there was not any sedimentation, only diffusion). 
I have also some question with respect to the calculation of coefficients, I would go through the source code and comment on specific lines

adimako commented:
@cekees:
1.) In Rans3PSed, why do we pass the viscosities. I'd imagine that the molecular is used for calculating the Darcy type drag, but I cannot find any use for the turbulent
2.) The SmoothedHS functions are derived from a phi value, is this phi reconstructed from the VOS
3.) the momentum advective fluxes (both explicit and implicit are 0, but is this related to the projection Scheme?
4.) As said, I would recommend putting the diffusion fluxes to zero for now. Later on, we can switch on the diffusion fluxes and get the viscosity from the collision / contact stress module
5.) The Hamiltonian terms I presume relate to the projection scheme. They seem fine, but I will check again

Let's discuss these at the meeting

cekees commented:
> 1.) In Rans3PSed, why do we pass the viscosities. I'd imagine that the molecular is used for calculating the Darcy type drag, but I cannot find any use for the turbulent

The eddy viscosity is just passed in because RANS3PF was the starting point. It can be removed.

> 2.) The SmoothedHS functions are derived from a phi value, is this phi reconstructed from the VOS

No, phi is for the air/water interface and based on the VOF. These are used to correctly set the fluid density and viscosity when needed, though presumably there will be no sediment in the air phase:)

> 3.) the momentum advective fluxes (both explicit and implicit are 0, but is this related to the projection Scheme?

Yes, it's because the momentum is switched to non-conservative form.

> 4.) As said, I would recommend putting the diffusion fluxes to zero for now. Later on, we can switch on the diffusion fluxes and get the viscosity from the collision / contact stress module

Sounds good.

> 5.) The Hamiltonian terms I presume relate to the projection scheme. They seem fine, but I will check again

Yes, thanks.

Giovanni-Cozzuto-1989 commented:
@cekees @manuel-quezada @adimako, I'm checking each term of the momentum equation in the solid phase. About the accumulation term, why do we miss the vos variable in it?
https://github.com/erdc/proteus/blob/gco_rans3p/proteus/mprans/RANS3PSed.h#L639

adimako commented:
@cekees @manuel-quezada 
![image](https://user-images.githubusercontent.com/8258298/31223892-a4491448-a9c3-11e7-8bc4-dd9206444a1e.png)

We have removed all diffusion and stress terms from the sediment equation, so the flow in the sediment should be boyancy and advection dominated. We still cannot get the sediment to settle. We cannot find anything wrong in the coefficient calculation, apart from the question that @Giovanni-Cozzuto-1989 posted above. 

One question I would have is about the sediment velocity, is it corrected in the projection scheme?

Giovanni-Cozzuto-1989 commented:
@cekees @adimako 
This is a first result (a bit unstable) and it's trying to do the right thing.
I switched off some terms and added vos in the accumulation term in momentum equation.
![immagine](https://user-images.githubusercontent.com/15638848/31394438-39dfa178-add6-11e7-95ca-68819fbd63f7.png)

Giovanni-Cozzuto-1989 commented:
@cekees @adimako @manuel-quezada 
We said the momentum equations are written in a non-conservative form. Why does RANS3PSed have all terms divided by density whilst RANS3PF doesn't? 
Should I set sediment's equation in a consistent way with fluid's?

Giovanni-Cozzuto-1989 commented:
Other thing to check is phi_s in RANS3PF and RANS3PSed:
- RANS3PF has porosity in it to take count of the sediment;
- RANS3PSed has phi_s then it's not clear why 1-phi_s is used frequently there.

Giovanni-Cozzuto-1989 commented:
Checking the correction equation terms for pressure and penalty coefficient used for perturbating Navier-Stokes.
https://github.com/erdc/proteus/blob/master/proteus/mprans/PresInc.h#L145

According to Guermond and Salgado (2008), the terms should be:
![immagine](https://user-images.githubusercontent.com/15638848/32277171-7be90f02-bf09-11e7-91ec-b298ea2f0be4.png)

cekees commented:
We moved that scaling to the other term so it’s the reciprocal, and it includes the changes for air/water/sediment.

Giovanni-Cozzuto-1989 commented:
Reasonable results with this latest version of the code.
Scaling of the pressure increment fixed.

Giovanni-Cozzuto-1989 commented on proteus/mprans/RANS3PSed2D.h:
```diff
@@ -828,41 +826,71 @@ namespace proteus
     }
 
+    inline
+        void updateParticlePressure(const double phi_s,
```
@adimako @cekees Could you check this part about frictional pressure in sediment phase?
What I mean is: should I add this term like it is?
The gradient is not needed because then we integrate through the cells, is it?

adimako replied:
Giovanni looks fine. 
I would remove the d_mom_source arrays completely. I do not think that there is a case where p_friction depends on the velocities

adimako commented on proteus/NumericalSolution.py:
```diff
@@ -1419,6 +1419,7 @@ def calculateSolution(self,runName):
+ 
                         #end model step
                         if stepFailed:
                             logEvent("Sequence step failed")
```
Ready to go

adimako commented:
Summary of additional work:

- [ ] Merged changes from gco

- [ ] Put condition for the penatly term, depending if the sediment phase is switched on. When it is switched on, the penatly term follows the formulation proposed by Giovanni

- [ ] Cleaned up / modified Giovannis file as much as I could to minimise differences with the master

- [ ] Fixed glitches in the boundary condition file

- [ ] Removed some repetivive code / bug

- [ ] Associated porosityTypes with porosity variable, not vos (although needs more work)

- [ ] Cleaned up and fixed holes in PresInc.py, with respect to definition and variable references

- [ ] Removed redundant reference of LS_model in CLSVOF (to be reviewed by @manuel-quezada )

- [ ] Relaxed tolerances in test_surf_tension from 1e-10 to 1e-7 to allow the tests to pass (to be reviewed by @manuel-quezada )

adimako commented:
@manuel-quezada are you also happy to relax the tolerances for passing the surface tension tests?

adimako commented:
@manuel-quezada I do not think that the Proj_EV tests will be affected but I understand your point, as there are significant changes in RANS3PF / Sed. 
Regarding the free-surface tension tests, I am not 100% confident know why the tolerances would be different and I saw that there is a comparison for all the elements of the array, so I thought maybe 1e-10 is too tight. But it might be not if e.g. the result is of that order.

manuel-quezada commented:
Those tests check the convergence of the whole projection scheme. A good number of things can affect the convergence rates. I tend to run them every time I make a major change but that is maybe me being picky. Anyway, if you didn't have to change the tolerance for the tests in `proteus/tests/ProjScheme_with_EV` then I am OK with the changes in the tolerances for the free-surface tension tests.

adimako commented:
@manuel-quezada I did not really interfere with the presInc.py (only to fix some bugs). I think some differences may come from the RANS3PF.py - this is why only the velocity comparison needs relaxing. @cekees can you have a look, I'd like to merge it so we can start running Giovanni's cases and maybe look at a couple more.

cekees commented on proteus/mprans/RANS3PF2D.h:
```diff
@@ -3720,7 +3717,7 @@ namespace proteus
                                                bc_dmom_u_ham_grad_p_ext[0],//=1/bc_rho,
                                                normal,
-                                               porosity_ext*ebqe_rho[ebNE_kb], // multiply by rho. mql. CHECK.
+                                               dmom_u_acc_u_ext,
```
@manuel-quezada I'm concerned about these changes to bc_mom_u_adv_ext, etc. and the line just above. Can you take a look as well? I don't believe these match changes to RANS3PF.h. These could explain the slightly larger errors in the test output.

cekees commented on proteus/mprans/RANS3PF2D.h:
```diff
@@ -5918,7 +5914,7 @@ namespace proteus
                                                           dmom_u_ham_grad_p_ext[0],//=1/rho
                                                           normal,
-                                                          porosity_ext*rhoSave, //multiply by rho. mql. CHECK.
+                                                          porosity_ext*dmom_u_acc_u_ext, //multiply by rho. mql. CHECK.
```
@manuel-quezada , these are similar to comment above.

cekees commented:
@adimako the lines I flagged above only affect the moving domain version of this. Unfortunately @manuel-quezada and I agree that there are still some issues with rans3p in a moving domain. This explains the results we got on @tridelat's caisson. So I'm merging this but the moving domain support will need a new PR.
